### PR TITLE
Update the JSON schema exporter to reuse schemas more aggressively.

### LINKED ISSUE #108755: OpenAPI hangs on XElement
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Describe the bug

Request or response with XElement in them causes OpenAPI generation to hang and allocate memory at a high rate.

### Expected Behavior

Should do something more sensible.

### Steps To Reproduce

Build and run the following, then load the OpenAPI json-file.
```c#
var builder = WebApplication.CreateBuilder(args);
builder.Services.AddOpenApi();

var app = builder.Build();
app.MapOpenApi();

app.MapGet("/", () => new XElement("Test")).WithOpenApi();

app.Run();
```

### Exceptions (if any)

We have instances of XElement in some of our contracts with a custom converter. With Swagger it was easy to map XElement to the schema for object which is appropriate for our use case.

This problem now occurs before any OpenApiSchemaTransformers are applied, as such we cannot get it to work at all.

### .NET Version

9.0.100-rc.1.24452.12

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
I'm on the fence whether this should be backported to .NET 9. On one hand it changes the shape of generated schemas which adds risk. On the other, the current memory blow-up experienced by generating schemas for `XElement` specifically seems to warrant servicing, even if the type itself is patently unsuitable for JSON serialization.
 
Fix #108755.

### CODE REVIEW

File: src/libraries/System.Text.Json/tests/Common/JsonSchemaExporterTests.cs [hunk 1]
```diff
@@ -90,6 +91,13 @@ public void UnsupportedType_ReturnsExpectedSchema(Type type)
             Assert.Equal(""""{"$comment":"Unsupported .NET type","not":true}"""", schema.ToJsonString());
         }
 
+        [Fact]
+        public void CanGenerateXElementSchema()
+        {
+            JsonNode schema = Serializer.DefaultOptions.GetJsonSchemaAsNode(typeof(XElement));
+            Assert.True(schema.ToJsonString().Length < 100_000);
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/tests/Common/JsonSchemaExporterTests.cs:98]: For context, the newly introduced schema reuse creates this monstrosity:

https://gist.github.com/eiriktsarpalis/fd53bb41b1f600a9b2e9c7985b9da64e

Comment (stephentoub) [reply src/libraries/System.Text.Json/tests/Common/JsonSchemaExporterTests.cs:98]: That's before or after this PR?

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/tests/Common/JsonSchemaExporterTests.cs:98]: After the PR. Before the PR it just blows up.

**Comment by jeffhandley** [comment]
With this change, are we just moving the line for what we'd fail with to be far enough out that we can handle `XElement`, but there will inevitably be some other type that we'd still fail on? What are the down-sides for this more aggressive reuse?

Without this change, is there another way one could get an `XElement` supported in their schema without the monstrosity schema that would be generated, or is that full schema truly required for that type/scenario?

**Comment by eiriktsarpalis** [comment]
> With this change, are we just moving the line for what we'd fail with to be far enough out that we can handle `XElement`, but there will inevitably be some other type that we'd still fail on?

The change should address the issue for all types, as the size of the generated schema is now bounded by the number types and properties present in the type graph.

> What are the down-sides for this more aggressive reuse?

More occurrences of `$ref` pointers in the generated schema. Semantically speaking the generated schemas are identical, but it should have negative impact from a human readability perspective.

> Without this change, is there another way one could get an XElement supported in their schema without the monstrosity schema that would be generated, or is that full schema truly required for that type/scenario?

Not that I'm aware of. FWIW passing `XElement` into STJ is for all intents and purposes not supported. This is not about fixing an otherwise valid use case, but rather it's making the library more resilient against random types that could be thrown at it.

**Comment by jeffhandley** [comment]
> The change should address the issue for all types, as the size of the generated schema is now bounded by the number types and properties present in the type graph.
> This is not about fixing an otherwise valid use case, but rather it's making the library more resilient against random types that could be thrown at it.

Thank you; that's helpful for me. With that context on the customer impact and risk, I support this change going in and being queued up for backport to release/9.0.

**Review by stephentoub — APPROVED** [review]

**Pull request merged**